src/Microsoft.AspNetCore.Sockets/WebSockets.cs

```diff
@@ -43,44 +45,61 @@ public WebSockets(Connection connection, Format format, ILoggerFactory loggerFac
+            var trigger = await Task.WhenAny(
+                receiving,
+                sending);
 
-                    // Send the server's close frame
-                    await ws.CloseAsync(WebSocketCloseStatus.NormalClosure);
-                }
-                else
+            // What happened?
+            if (trigger == receiving)
+            {
+                if (receiving.IsCanceled || receiving.IsFaulted)
                 {
-                    // The application finished sending. We're not going to keep the connection open,
-                    // so close it and wait for the client to ack the close
-                    _channel.Input.CompleteWriter();
-                    _logger.LogDebug("Application finished sending. Sending close frame.");
-                    await ws.CloseAsync(WebSocketCloseStatus.NormalClosure);
-
-                    _logger.LogDebug("Waiting for the client to close the socket");
-                    // TODO: Timeout.
-                    await receiving;
+                    // The receiver faulted or cancelled. This means the client is probably broken. Just propagate the exception and exit
+                    receiving.GetAwaiter().GetResult();
+
+                    // Should never get here because GetResult above will throw
+                    Debug.Fail("GetResult didn't throw?");
+                    return;
                 }
+
+                // Shutting down because we received a close frame from the client.
+                // Complete the input writer so that the application knows there won't be any more input.
+                _logger.LogDebug("Client closed connection with status code '{0}' ({1}). Signaling end-of-input to application", receiving.Result.Status, receiving.Result.Description);
+                _channel.Input.CompleteWriter();
+
+                // Wait for the application to finish sending.
+                _logger.LogDebug("Waiting for the application to finish sending data");
+                await sending;
+
+                // Send the server's close frame
+                await socket.CloseAsync(WebSocketCloseStatus.NormalClosure);
+            }
+            else
+            {
+                var failed = sending.IsFaulted || sending.IsCompleted;
+                // The application finished sending. We're not going to keep the connection open,
+                // so close it and wait for the client to ack the close
+                _channel.Input.CompleteWriter();
+                _logger.LogDebug(!failed ? "Application finished sending. Sending close frame." : "Application failed during sending. Sending InternalServerError close frame");
+                await socket.CloseAsync(!failed ? WebSocketCloseStatus.NormalClosure : WebSocketCloseStatus.InternalServerError);
+
+                _logger.LogDebug("Waiting for the client to close the socket");
+
+                // TODO: Timeout.
+                await receiving;
```

Comment:
I need to think about this. I had this comment before:

https://github.com/aspnet/SignalR/blob/d00f1f93b2676706203f0ce65251a68bea9cfcbe/src/Microsoft.AspNetCore.Sockets/WebSockets.cs#L35-L38